File: tests/io/file_pipe.js
function pipe_open() {
  const ffi = require("bun:ffi");
  const lib = ffi.dlopen(process.platform === "darwin" ? "libSystem.dylib" : "libc.so.6",
    { pipe: { args: ["ptr"], returns: "i32" } });
  const fd = new Int32Array(2);
  const n = lib.symbols.pipe(ffi.ptr(fd));
  const code = n < 0 ? io_sys().errno() : 0;
  lib.close();
  if (n < 0) return io_fail(code);
  const fs = require("fs");
  try {
    fs.writeSync(fd[1], "abc");
    return io_done(fd[0]);
  } catch (e) {
    fs.closeSync(fd[0]);
    return io_fail(-e.errno);
  } finally {
    fs.closeSync(fd[1]);
  }
}

File: tests/io/file_pipe.c
Term pipe_open_run(Env e, Term* f, IoWork* w) {
  int fd[2];
  if (pipe(fd) < 0) return io_fail(e, errno, NULL);
  ssize_t n = write(fd[1], "abc", 3);
  int code = n < 0 ? errno : EIO;
  close(fd[1]);
  if (n != 3) {
    close(fd[0]);
    return io_fail(e, code, NULL);
  }
  return io_done(e, io_hand(fd[0]));
}

static void __attribute__((constructor)) pipe_open_use(void) {
  io_eff(CID_PIPE_OPEN, pipe_open_run, 0);
}
